lib/tarfiles.py

```diff
@@ -128,50 +145,132 @@ def do_tarballs(args):
-    url = "https://%s/pubapi/publish?cid=%s" % (dropbox_server, cid)
-    if tokenstr:
-        headers = { 'Authorization': 'Bearer %s' % tokenstr}
-        res = requests.post(url, headers=headers, data=tf, verify=False)
-    else:
-        res = requests.post(url, cert=(proxy, proxy), data=tf, verify=False)
-    if res.text[:8] == "PRESENT:":
-        return res.text[8:]
-    else:
-        return None
-
-
-def pubapi_exists(cid, proxy = None, tokenstr = None):
-    """make pubapi update call to check if we already have this tarfile,
-    return path.
+class TarfilePublisherHandler(object):
+    """Handler to publish tarballs via HTTP to RCDS (or future dropbox server)
+
+    Args:
+        object (_type_): _description_
+        cid (str): unique group/hash combination that RCDS uses to locate tarballs
+        proxy (str): Location of X509 Proxy file to authenticate to RCDS
+        token (str): Location of JWT/Sci-token to authenticate to RCDS
     """
-    dropbox_server = "rcds.fnal.gov"
-    url = "https://%s/pubapi/exists?cid=%s" % (dropbox_server, cid)
-    if tokenstr:
-        headers = { 'Authorization': 'Bearer %s' % tokenstr}
-        res = requests.get(url, headers=headers , verify=False)
-    else:
-        res = requests.get(url, cert=(proxy, proxy), verify=False)
-    if res.text[:8] == "PRESENT:":
-        return res.text[8:]
-    else:
-        return None
+    dropbox_server_string = os.getenv("JOBSUB_DROPBOX_SERVER_LIST", "rcds01.fnal.gov rcds02.fnal.gov")
+    check_tarball_present_re = re.compile(
+        "^PRESENT\:(.+)$"
+    )  # RCDS returns this if a tarball represented by cid is present
+
+    def __init__(self, cid, proxy=None, token=None):
+        self.cid_url = _quote(cid, safe="")  # Encode CID for passing to URL
+        self.proxy = proxy
+        self.token = token
+        if token is not None:
+            self.request_headers = self.__make_request_token_headers()
+            print(f"Using bearer token located at {self.token} to authenticate to RCDS")
+        else:
+            print(f"Using X509 proxy located at {self.proxy} to authenticate to RCDS")
+        self.dropbox_servers = tuple(self.dropbox_server_string.split())
+        self.pubapi_base_url_formatter = (
+            f"https://{{dropbox_server}}/pubapi/{{endpoint}}?cid={self.cid_url}"
+        )
+
+    # Some sort of wrapper to wrap the above three
+    def pubapi_operation(func):
+        """Wrap various PubAPI operations, return path if we get it from response"""
+
+        class SafeDict(dict):
+            """Use this object to allow us to not need all keys of dict when
+               running str.format_map method to do string interpolation.
+               Taken from https://stackoverflow.com/a/17215533"""
+            def __missing__(self, key):
+                return f"{{{key}}}" # "{<key>}"
+
+        def wrapper(self, *args, **kwargs):
+            _dropbox_server_selector = self.__select_dropbox_server()
```

Comment:
If I'm working through this right, we'll be trying a new server with each retry, rather than retrying a single server more than once? If that is what's desired, then is there any need for `NUM_RETRIES`? And for that matter any need for a delay between retries. IMO we should try a server more than once before moving on.

I'd also like to see a message (debug maybe?) saying what dropbox server is being tried.

Reply:
You're right - that's exactly how I wrote this:  So that we try a new server with each retry.  Is your thought that we should retry a single server multiple times, and if we exhaust `NUM_RETRIES`, then switch to a new server?

Reply:
That is my preference, yes, but it's worth some discussion. It would be simpler to just run through the server list, in which case you could get rid of the retry limit (now implicit in the number of servers) and maybe the interval (since you're hitting different servers, although it may help in case of transient local network issues, for example). 

I suppose in the case of only one server one could just list it multiple times...

Reply:
I'm not too married to my approach here - the way I wrote it guarantees that we hit a different server within 2 retries, but I think it's just as valid to retry a single server, then if it fails, pop it off the list and try the other for the retry limit.  I'd rather keep the retry limit, because of exactly what you said in your last point.  If there's a transient local network issue, and we have a cluster of, say, 2 servers, we'd only retry once, which may not be enough.  

Doing it like you first suggested, where we try one server for `NUM_RETRIES + 1` tries, then switch to the next one, and so on, takes longer before it fails, but it allows for some time in case of a transient issue, while ensuring that we tried each server multiple times.  So if you're OK with it, I'll keep the retry limit and interval in there, but switch the logic so we try the servers the way you suggested.